keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -161,273 +122,289 @@ consistent with the syntax supported by any of the common proxy vendors:
+- `"*.foo.com"` does not match `"aaa.bbb.foo.com"` as the wildcard only
+   matches a single label.
+- `"*.foo.com"` does not match `"foo.com"`, as the wildcard must match a
+   single label.
+
+Note: label refers to a "DNS label", i.e. the strings separated by the dots "."
+in the domain name.
 
 ### Status
 
-!!! TODO
+As this is strictly additive, this could be punted to post-GA to reduce
+the size of the change.
 
 ### Ingress class
 
-The `kubernetes.io/ingress.class` annotation is required for interoperation
-between various Ingress providers. As support for this annotation is universal,
-this concept should be promoted to an actual field.
+The `kubernetes.io/ingress.class` annotation is required for selecting between
+multiple Ingress providers. As support for this annotation is universal, this
+concept should be promoted to an actual field.
 
-#### Proposal
+#### Ingress class proposal
 
-Promoting the annotation as is as an opaque string the most direct path, but
-precludes any future enhancements to the concept.
+Promoting the annotation as it is currently defined as an opaque string is the
+most direct path but precludes any future enhancements to the concept.
 
 An alternative is to create a new resource `IngressClass` to take its place.
 This resource will serve a couple of purposes:
 
-* Define the set of valid classes available to the user. Gives operators control
+- Define the set of valid classes available to the user. Gives operators control
   over allowed classes.
-* Allow us to evolve the API to express concepts such a levels of service
+- Allow us to evolve the API to express concepts such a levels of service
   associated with a given Ingress controller.
 
 Add a field to `ingress.spec`:
+
 ```golang
 type IngressSpec struct {
   ...
-  // Class references an IngressClass resource in kube-system. This is used
-  // by the cluster Ingress controllers to determine which controller operates
-  // on this resource.
+  // Class references an IngressClass cluster resource. This is used
+  // by the cluster Ingress controllers to determine which controller
+  // operates on this resource.
   Class string
   ...
 }
 ```
 
 ```golang
 // IngressClass represents the class of the Ingress, referenced by the
-// ingress.spec.
+// ingress.spec. IngressClass will be a non-namespaced Cluster resource.
 type IngressClass struct {
   metav1.TypeMeta
   metav1.ObjectMeta
-  Spec IngressClassSpec
-}
-
-type IngressClassSpec struct {
-  // This is currently empty.
 }
 ```
 
+The `IngressClass` object will be left empty for fields for the time
+being, but can be extended using annotations. It is likely we will
+find common fields appropriate for the resource as this gains
+adoption.
+
+##### Interoperability with previous annotation
+
+TODO: We have a choice here: roundtrip with the annotation to class
+(e.g. take the annotation value and make it the `Class` string. Or
+support both values and prioritize the annotation as the class.
```

Comment:
There's no standard form of ingress.class values, right?  It certainly does not mean "the IngressClass instance with the name ...".

I think we have to say that the annotation takes precedence, and the "new" class is totally ignored if the annotation is present.  Maybe even validate the XOR of them

Reply:
Added comment about prioritizing annotation, supporting both.